### ISSUE TITLE
Constant "Deadlock detected and averted, waiting" message during Stage 2

### ISSUE THREAD

**Issue by DanielJamieson** [findie/wikidata-neo4j-importer#10]
Having a problem running stage 2 at the moment. It tries linking entities and just gives me a message "Deadlock detected and averted". Below is my configuration file:

> {
>   "file": "latest-all.json",
>   "neo4j": {
>     "bolt": "bolt://localhost",
>     "auth": {
>       "user": "neo4j",
>       "pass": "neo4j_password"
>     }
>   },
>   "do": {
>     "0": false,
>     "1": false,
>     "2": true,
>     "3": true
>   },
>   "verbose": false,
>   "concurrency": 4,
>   "skip": 0,
>   "lines": 54899326,
>   "bucket": 1000
> }

The VPS has 8GB of RAM and 4 CPUs. I can monitor the RAM and CPU and it's not maxing out at all.

I've been running the project with this command (so that I can logout of my VPS and then log back in again): `nohup node index.js &`

And this is the output:

> Counting lines
> Found 54899326 lines
> Starting simple node relationship creation...
> Imported 0 lines
> 0% done!  ->  Remaining 
> Imported 1000 lines
> 0.001% done!  ->  Remaining 19h 58m 4s 285ms
> Imported 2000 lines
> 0.003% done!  ->  Remaining 1h 37m 14s 381ms
> Imported 3000 lines
> 0.005% done!  ->  Remaining 4h 36m 39s 141ms
> Linking entities (2)                                                  : 100866.640ms
> Imported 4000 lines
> 0.007% done!  ->  Remaining 12h 6m 50s 640ms
> Linking entities (3)                                                  : 105860.533ms
> Deadlock detected and averted, waiting 716ms!
> Deadlock detected and averted, waiting 643ms!
> Deadlock detected and averted, waiting 781ms!
> Deadlock detected and averted, waiting 928ms!
> Deadlock detected and averted, waiting 524ms!
> Deadlock detected and averted, waiting 525ms!
> Deadlock detected and averted, waiting 985ms!
> Deadlock detected and averted, waiting 757ms!
> Deadlock detected and averted, waiting 997ms!
> Deadlock detected and averted, waiting 730ms!
> Linking entities (1)                                                  : 112289.142ms
> Deadlock detected and averted, waiting 826ms!
> Deadlock detected and averted, waiting 823ms!
> Deadlock detected and averted, waiting 619ms!
> Deadlock detected and averted, waiting 799ms!
> Deadlock detected and averted, waiting 550ms!
> Deadlock detected and averted, waiting 860ms!
> Deadlock detected and averted, waiting 703ms!
> Deadlock detected and averted, waiting 770ms!
> Deadlock detected and averted, waiting 877ms!
> Deadlock detected and averted, waiting 985ms!
> Deadlock detected and averted, waiting 530ms!
> Deadlock detected and averted, waiting 940ms!
> Deadlock detected and averted, waiting 842ms!
> Deadlock detected and averted, waiting 617ms!
> Deadlock detected and averted, waiting 609ms!
> Deadlock detected and averted, waiting 527ms!
> Deadlock detected and averted, waiting 853ms!
> Deadlock detected and averted, waiting 897ms!
> Deadlock detected and averted, waiting 514ms!
> Deadlock detected and averted, waiting 827ms!
> Deadlock detected and averted, waiting 709ms!
> Deadlock detected and averted, waiting 852ms!
> Deadlock detected and averted, waiting 855ms!
> Linking entities (0)                                                  : 121785.154ms
> Deadlock detected and averted, waiting 983ms!
> Deadlock detected and averted, waiting 514ms!
> Deadlock detected and averted, waiting 616ms!
> Deadlock detected and averted, waiting 537ms!
> Deadlock detected and averted, waiting 779ms!
> Deadlock detected and averted, waiting 691ms!
> Deadlock detected and averted, waiting 510ms!
> Deadlock detected and averted, waiting 599ms!
> Deadlock detected and averted, waiting 571ms!
> Deadlock detected and averted, waiting 694ms!
> Deadlock detected and averted, waiting 743ms!
> Deadlock detected and averted, waiting 968ms!
> Deadlock detected and averted, waiting 736ms!
> Deadlock detected and averted, waiting 520ms!
> Deadlock detected and averted, waiting 779ms!
> Deadlock detected and averted, waiting 522ms!
> Deadlock detected and averted, waiting 859ms!
> Deadlock detected and averted, waiting 750ms!
> Deadlock detected and averted, waiting 615ms!
> Deadlock detected and averted, waiting 948ms!
> Deadlock detected and averted, waiting 527ms!
> Deadlock detected and averted, waiting 680ms!
> Deadlock detected and averted, waiting 732ms!
> Deadlock detected and averted, waiting 913ms!
> Deadlock detected and averted, waiting 720ms!
> Deadlock detected and averted, waiting 895ms!
> Deadlock detected and averted, waiting 735ms!
> Deadlock detected and averted, waiting 724ms!
> Deadlock detected and averted, waiting 649ms!
> Deadlock detected and averted, waiting 917ms!
> Deadlock detected and averted, waiting 963ms!
> Deadlock detected and averted, waiting 509ms!
> Deadlock detected and averted, waiting 783ms!
> Deadlock detected and averted, waiting 610ms!
> Deadlock detected and averted, waiting 728ms!
> Deadlock detected and averted, waiting 988ms!
> Deadlock detected and averted, waiting 893ms!
> Deadlock detected and averted, waiting 910ms!
> Deadlock detected and averted, waiting 951ms!
> Deadlock detected and averted, waiting 591ms!
> Deadlock detected and averted, waiting 610ms!
> Deadlock detected and averted, waiting 828ms!
> Deadlock detected and averted, waiting 654ms!
> Deadlock detected and averted, waiting 788ms!
> Deadlock detected and averted, waiting 977ms!
> Deadlock detected and averted, waiting 756ms!
> Deadlock detected and averted, waiting 623ms!
> Deadlock detected and averted, waiting 959ms!
> Deadlock detected and averted, waiting 668ms!
> Deadlock detected and averted, waiting 930ms!
> Deadlock detected and averted, waiting 743ms!
> Deadlock detected and averted, waiting 711ms!
> Deadlock detected and averted, waiting 988ms!
> Deadlock detected and averted, waiting 997ms!
> Deadlock detected and averted, waiting 600ms!
> Deadlock detected and averted, waiting 523ms!
> Deadlock detected and averted, waiting 721ms!
> Deadlock detected and averted, waiting 569ms!
> Deadlock detected and averted, waiting 822ms!
> Deadlock detected and averted, waiting 783ms!
> Deadlock detected and averted, waiting 755ms!
> Deadlock detected and averted, waiting 838ms!
> Deadlock detected and averted, waiting 851ms!
> Deadlock detected and averted, waiting 517ms!
> Deadlock detected and averted, waiting 656ms!
> Deadlock detected and averted, waiting 957ms!
> Deadlock detected and averted, waiting 966ms!
> Deadlock detected and averted, waiting 750ms!
> Deadlock detected and averted, waiting 599ms!
> Deadlock detected and averted, waiting 884ms!
> Deadlock detected and averted, waiting 707ms!
> Deadlock detected and averted, waiting 835ms!
> Deadlock detected and averted, waiting 519ms!
> Deadlock detected and averted, waiting 705ms!
> Deadlock detected and averted, waiting 560ms!
> Deadlock detected and averted, waiting 565ms!
> Deadlock detected and averted, waiting 987ms!
> Deadlock detected and averted, waiting 725ms!
> Deadlock detected and averted, waiting 917ms!
> Deadlock detected and averted, waiting 716ms!
> Deadlock detected and averted, waiting 828ms!
> Deadlock detected and averted, waiting 741ms!
> Deadlock detected and averted, waiting 581ms!
> Deadlock detected and averted, waiting 627ms!
> Deadlock detected and averted, waiting 984ms!
> Deadlock detected and averted, waiting 691ms!
> Deadlock detected and averted, waiting 629ms!
> Deadlock detected and averted, waiting 662ms!
> Deadlock detected and averted, waiting 667ms!
> Deadlock detected and averted, waiting 903ms!
> Deadlock detected and averted, waiting 743ms!
> Deadlock detected and averted, waiting 894ms!
> Deadlock detected and averted, waiting 538ms!
> Deadlock detected and averted, waiting 694ms!
> Deadlock detected and averted, waiting 549ms!
> Deadlock detected and averted, waiting 590ms!
> Deadlock detected and averted, waiting 930ms!
> Deadlock detected and averted, waiting 611ms!
> Deadlock detected and averted, waiting 701ms!
> Deadlock detected and averted, waiting 905ms!
> Deadlock detected and averted, waiting 808ms!
> Deadlock detected and averted, waiting 928ms!
> Deadlock detected and averted, waiting 594ms!
> Deadlock detected and averted, waiting 633ms!
> Deadlock detected and averted, waiting 861ms!
> Deadlock detected and averted, waiting 867ms!
> Deadlock detected and averted, waiting 605ms!
> Deadlock detected and averted, waiting 984ms!
> Deadlock detected and averted, waiting 839ms!
> Deadlock detected and averted, waiting 639ms!
> Deadlock detected and averted, waiting 823ms!
> Deadlock detected and averted, waiting 750ms!
> Deadlock detected and averted, waiting 593ms!
> Deadlock detected and averted, waiting 560ms!
> Deadlock detected and averted, waiting 942ms!
> Deadlock detected and averted, waiting 606ms!
> Deadlock detected and averted, waiting 650ms!
> Deadlock detected and averted, waiting 828ms!
> Deadlock detected and averted, waiting 876ms!
> Deadlock detected and averted, waiting 813ms!
> Deadlock detected and averted, waiting 868ms!
> Deadlock detected and averted, waiting 567ms!
> Deadlock detected and averted, waiting 631ms!
> Deadlock detected and averted, waiting 963ms!
> Deadlock detected and averted, waiting 699ms!
> Deadlock detected and averted, waiting 740ms!
> Deadlock detected and averted, waiting 653ms!
> Deadlock detected and averted, waiting 897ms!
> Deadlock detected and averted, waiting 510ms!
> Deadlock detected and averted, waiting 871ms!
> Deadlock detected and averted, waiting 982ms!
> Deadlock detected and averted, waiting 747ms!
> Deadlock detected and averted, waiting 797ms!
> Deadlock detected and averted, waiting 824ms!
> Deadlock detected and averted, waiting 824ms!
> Deadlock detected and averted, waiting 612ms!
> Deadlock detected and averted, waiting 716ms!
> Deadlock detected and averted, waiting 550ms!
> Deadlock detected and averted, waiting 928ms!
> Deadlock detected and averted, waiting 551ms!
> Deadlock detected and averted, waiting 909ms!
> Deadlock detected and averted, waiting 537ms!
> Deadlock detected and averted, waiting 524ms!
> Deadlock detected and averted, waiting 549ms!
> Deadlock detected and averted, waiting 771ms!
> Deadlock detected and averted, waiting 897ms!
> Deadlock detected and averted, waiting 502ms!
> Deadlock detected and averted, waiting 873ms!
> Deadlock detected and averted, waiting 599ms!
> Deadlock detected and averted, waiting 724ms!
> Deadlock detected and averted, waiting 797ms!
> Deadlock detected and averted, waiting 821ms!
> Deadlock detected and averted, waiting 990ms!
> Deadlock detected and averted, waiting 523ms!
> Deadlock detected and averted, waiting 613ms!
> Deadlock detected and averted, waiting 582ms!
> Deadlock detected and averted, waiting 931ms!
> Deadlock detected and averted, waiting 710ms!
> Deadlock detected and averted, waiting 934ms!
> Deadlock detected and averted, waiting 629ms!
> Deadlock detected and averted, waiting 900ms!
> Deadlock detected and averted, waiting 948ms!
> Deadlock detected and averted, waiting 917ms!
> Deadlock detected and averted, waiting 635ms!
> Deadlock detected and averted, waiting 771ms!
> Deadlock detected and averted, waiting 728ms!
> Deadlock detected and averted, waiting 565ms!
> Deadlock detected and averted, waiting 751ms!
> Deadlock detected and averted, waiting 703ms!
> Deadlock detected and averted, waiting 688ms!
> Deadlock detected and averted, waiting 954ms!
> Deadlock detected and averted, waiting 816ms!
> Deadlock detected and averted, waiting 865ms!
> Deadlock detected and averted, waiting 653ms!
> Deadlock detected and averted, waiting 522ms!
> Deadlock detected and averted, waiting 568ms!
> Deadlock detected and averted, waiting 674ms!
> Deadlock detected and averted, waiting 808ms!
> Deadlock detected and averted, waiting 722ms!
> Deadlock detected and averted, waiting 654ms!
> Deadlock detected and averted, waiting 612ms!
> Deadlock detected and averted, waiting 629ms!
> Deadlock detected and averted, waiting 703ms!
> Deadlock detected and averted, waiting 990ms!
> Deadlock detected and averted, waiting 628ms!
> Deadlock detected and averted, waiting 730ms!
> Deadlock detected and averted, waiting 714ms!
> Deadlock detected and averted, waiting 994ms!
> Deadlock detected and averted, waiting 874ms!
> Deadlock detected and averted, waiting 792ms!
> Deadlock detected and averted, waiting 876ms!
> Deadlock detected and averted, waiting 810ms!
> Deadlock detected and averted, waiting 626ms!
> Deadlock detected and averted, waiting 812ms!
> Deadlock detected and averted, waiting 825ms!
> Deadlock detected and averted, waiting 621ms!
> Deadlock detected and averted, waiting 504ms!
> Deadlock detected and averted, waiting 764ms!
> Deadlock detected and averted, waiting 558ms!
> Deadlock detected and averted, waiting 613ms!
> Deadlock detected and averted, waiting 854ms!
> Deadlock detected and averted, waiting 604ms!
> Deadlock detected and averted, waiting 908ms!
> Deadlock detected and averted, waiting 724ms!
> Deadlock detected and averted, waiting 640ms!
> Deadlock detected and averted, waiting 966ms!
> Deadlock detected and averted, waiting 988ms!
> Deadlock detected and averted, waiting 735ms!
> Deadlock detected and averted, waiting 895ms!
> Deadlock detected and averted, waiting 785ms!
> Deadlock detected and averted, waiting 807ms!
> Deadlock detected and averted, waiting 583ms!
> Deadlock detected and averted, waiting 732ms!
> Deadlock detected and averted, waiting 981ms!
> Deadlock detected and averted, waiting 670ms!
> Deadlock detected and averted, waiting 798ms!
> Deadlock detected and averted, waiting 991ms!
> Deadlock detected and averted, waiting 941ms!
> Deadlock detected and averted, waiting 717ms!
> Deadlock detected and averted, waiting 592ms!
> Deadlock detected and averted, waiting 680ms!
> Deadlock detected and averted, waiting 532ms!
> Deadlock detected and averted, waiting 519ms!
> Deadlock detected and averted, waiting 530ms!
> Deadlock detected and averted, waiting 775ms!
> Deadlock detected and averted, waiting 660ms!
> Deadlock detected and averted, waiting 896ms!
> Deadlock detected and averted, waiting 626ms!
> Deadlock detected and averted, waiting 955ms!
> Deadlock detected and averted, waiting 568ms!
> Deadlock detected and averted, waiting 715ms!
> Deadlock detected and averted, waiting 569ms!
> Deadlock detected and averted, waiting 814ms!
> Deadlock detected and averted, waiting 741ms!
> Deadlock detected and averted, waiting 650ms!
> Deadlock detected and averted, waiting 580ms!
> Deadlock detected and averted, waiting 640ms!
> Deadlock detected and averted, waiting 682ms!
> Deadlock detected and averted, waiting 523ms!
> Deadlock detected and averted, waiting 955ms!
> Deadlock detected and averted, waiting 838ms!
> Deadlock detected and averted, waiting 743ms!
> Deadlock detected and averted, waiting 930ms!
> Deadlock detected and averted, waiting 669ms!
> Deadlock detected and averted, waiting 606ms!
> Deadlock detected and averted, waiting 971ms!
> Deadlock detected and averted, waiting 804ms!
> Deadlock detected and averted, waiting 687ms!
> Deadlock detected and averted, waiting 839ms!
> Deadlock detected and averted, waiting 787ms!
> Deadlock detected and averted, waiting 647ms!
> Deadlock detected and averted, waiting 704ms!
> Deadlock detected and averted, waiting 982ms!
> Deadlock detected and averted, waiting 662ms!
> Deadlock detected and averted, waiting 649ms!
> Deadlock detected and averted, waiting 986ms!
> Deadlock detected and averted, waiting 537ms!
> Deadlock detected and averted, waiting 689ms!
> Deadlock detected and averted, waiting 672ms!
> Deadlock detected and averted, waiting 627ms!
> Deadlock detected and averted, waiting 781ms!
> Deadlock detected and averted, waiting 647ms!
> Deadlock detected and averted, waiting 775ms!
> Deadlock detected and averted, waiting 929ms!
> Deadlock detected and averted, waiting 945ms!
> Deadlock detected and averted, waiting 763ms!
> Deadlock detected and averted, waiting 876ms!
> Deadlock detected and averted, waiting 563ms!
> Deadlock detected and averted, waiting 849ms!
> Deadlock detected and averted, waiting 932ms!
> Deadlock detected and averted, waiting 939ms!
> Deadlock detected and averted, waiting 735ms!
> Deadlock detected and averted, waiting 666ms!
> Deadlock detected and averted, waiting 714ms!
> Deadlock detected and averted, waiting 959ms!
> Deadlock detected and averted, waiting 782ms!
> Deadlock detected and averted, waiting 558ms!
> Deadlock detected and averted, waiting 838ms!
> Deadlock detected and averted, waiting 540ms!
> Deadlock detected and averted, waiting 600ms!
> Deadlock detected and averted, waiting 965ms!
> Deadlock detected and averted, waiting 788ms!
> Deadlock detected and averted, waiting 953ms!
> Deadlock detected and averted, waiting 718ms!
> Deadlock detected and averted, waiting 538ms!
> Deadlock detected and averted, waiting 695ms!
> Deadlock detected and averted, waiting 933ms!
> Deadlock detected and averted, waiting 766ms!
> Deadlock detected and averted, waiting 674ms!
> Deadlock detected and averted, waiting 529ms!
> Deadlock detected and averted, waiting 865ms!
> Deadlock detected and averted, waiting 786ms!
> Deadlock detected and averted, waiting 751ms!
> Deadlock detected and averted, waiting 721ms!
> Deadlock detected and averted, waiting 556ms!
> Deadlock detected and averted, waiting 836ms!
> Deadlock detected and averted, waiting 644ms!
> Deadlock detected and averted, waiting 564ms!
> Deadlock detected and averted, waiting 805ms!
> Deadlock detected and averted, waiting 743ms!
> Deadlock detected and averted, waiting 660ms!
> Deadlock detected and averted, waiting 846ms!
> Deadlock detected and averted, waiting 977ms!
> Deadlock detected and averted, waiting 601ms!
> Deadlock detected and averted, waiting 877ms!
> Deadlock detected and averted, waiting 929ms!
> Deadlock detected and averted, waiting 603ms!
> Deadlock detected and averted, waiting 500ms!
> Deadlock detected and averted, waiting 849ms!
> Deadlock detected and averted, waiting 891ms!
> Deadlock detected and averted, waiting 970ms!
> Deadlock detected and averted, waiting 865ms!
> Deadlock detected and averted, waiting 612ms!
> Deadlock detected and averted, waiting 886ms!
> Deadlock detected and averted, waiting 865ms!
> Deadlock detected and averted, waiting 819ms!
> Deadlock detected and averted, waiting 514ms!
> Deadlock detected and averted, waiting 763ms!
> Deadlock detected and averted, waiting 906ms!
> Deadlock detected and averted, waiting 524ms!
> Deadlock detected and averted, waiting 909ms!
> Deadlock detected and averted, waiting 604ms!
> Deadlock detected and averted, waiting 667ms!
> Deadlock detected and averted, waiting 668ms!
> Deadlock detected and averted, waiting 737ms!
> Deadlock detected and averted, waiting 623ms!
> Deadlock detected and averted, waiting 689ms!
> Deadlock detected and averted, waiting 952ms!
> Deadlock detected and averted, waiting 788ms!
> Deadlock detected and averted, waiting 800ms!
> Deadlock detected and averted, waiting 776ms!
> Deadlock detected and averted, waiting 671ms!
> Deadlock detected and averted, waiting 957ms!
> Deadlock detected and averted, waiting 932ms!
> Deadlock detected and averted, waiting 738ms!
> Deadlock detected and averted, waiting 512ms!
> Deadlock detected and averted, waiting 517ms!
> Deadlock detected and averted, waiting 551ms!
> Deadlock detected and averted, waiting 822ms!
> Deadlock detected and averted, waiting 929ms!
> Deadlock detected and averted, waiting 738ms!
> Deadlock detected and averted, waiting 566ms!
> Deadlock detected and averted, waiting 656ms!
> Deadlock detected and averted, waiting 501ms!
> Deadlock detected and averted, waiting 562ms!
> Deadlock detected and averted, waiting 709ms!
> Deadlock detected and averted, waiting 850ms!
> Deadlock detected and averted, waiting 633ms!
> Deadlock detected and averted, waiting 952ms!
> Deadlock detected and averted, waiting 896ms!
> Deadlock detected and averted, waiting 696ms!
> Deadlock detected and averted, waiting 770ms!
> Deadlock detected and averted, waiting 903ms!
> Deadlock detected and averted, waiting 910ms!
> Deadlock detected and averted, waiting 821ms!
> Deadlock detected and averted, waiting 907ms!
> Deadlock detected and averted, waiting 746ms!
> Deadlock detected and averted, waiting 938ms!
> Deadlock detected and averted, waiting 788ms!
> Deadlock detected and averted, waiting 553ms!
> Deadlock detected and averted, waiting 753ms!
> Deadlock detected and averted, waiting 853ms!
> Deadlock detected and averted, waiting 985ms!
> Deadlock detected and averted, waiting 991ms!
> Deadlock detected and averted, waiting 716ms!
> Deadlock detected and averted, waiting 801ms!
> Deadlock detected and averted, waiting 847ms!
> Deadlock detected and averted, waiting 987ms!
> Deadlock detected and averted, waiting 691ms!
> Deadlock detected and averted, waiting 506ms!
> Deadlock detected and averted, waiting 570ms!
> Deadlock detected and averted, waiting 966ms!
> Deadlock detected and averted, waiting 702ms!
> Deadlock detected and averted, waiting 696ms!
> Deadlock detected and averted, waiting 888ms!
> Deadlock detected and averted, waiting 741ms!
> Deadlock detected and averted, waiting 985ms!
> Deadlock detected and averted, waiting 526ms!
> Linking entities (2)                                                  : 105330.665ms
> Deadlock detected and averted, waiting 964ms!
> Deadlock detected and averted, waiting 648ms!
> Deadlock detected and averted, waiting 733ms!
> Deadlock detected and averted, waiting 696ms!
> Deadlock detected and averted, waiting 806ms!
> Deadlock detected and averted, waiting 789ms!
> Deadlock detected and averted, waiting 907ms!
> Deadlock detected and averted, waiting 992ms!
> Deadlock detected and averted, waiting 708ms!
> Deadlock detected and averted, waiting 770ms!
> Deadlock detected and averted, waiting 828ms!
> Deadlock detected and averted, waiting 691ms!
> Deadlock detected and averted, waiting 513ms!
> Deadlock detected and averted, waiting 584ms!
> Deadlock detected and averted, waiting 775ms!
> Deadlock detected and averted, waiting 966ms!
> Deadlock detected and averted, waiting 577ms!
> Deadlock detected and averted, waiting 605ms!
> Deadlock detected and averted, waiting 683ms!
> Deadlock detected and averted, waiting 912ms!
> Deadlock detected and averted, waiting 988ms!
> Deadlock detected and averted, waiting 622ms!
> Deadlock detected and averted, waiting 747ms!
> Deadlock detected and averted, waiting 883ms!
> Deadlock detected and averted, waiting 633ms!
> Deadlock detected and averted, waiting 720ms!
> Deadlock detected and averted, waiting 734ms!
> Deadlock detected and averted, waiting 652ms!
> Deadlock detected and averted, waiting 556ms!

And it goes on and on like that with the deadlock line. 

Finally, there are no complaints whatsoever in the neo4j log.

What could be causing this issue?

**Comment by legraphista** [comment]
It looks like there are competing queries that want to modify the same subgraph. 
This can be caused by either a slow graph update or high a concurrency setting.
Try setting the concurrency to 1, it will slow down the process but should fix the issue.

**Comment by DanielJamieson** [comment]
Hey @legraphista, thank you for replying to me!

Unfortunately I already tried that. I have tried with concurrency = 4 and with concurrency = 1.

I have even tried to lower the bucket size from 1000 to 1, and neither of those settings make any difference whatsoever. 

Basically what happens is that it runs into this deadlock issue, that happens over and over again just like in the logs above - and eventually the script just gives up.

I have not been able to finish this Stage 2 at all yet. 

Could it have anything to do with the JSON being updated over time? I notice this project started back in 2017 and the readme states that it's ~120GB for the wikidata JSON file, but mine is 682GB, which means a lot has been added since then.

If it's not that and it is due to a slow graph update, I could try upgrading the machine although that would be a last resort.

**Comment by legraphista** [comment]
As you pointed out, when the project started back in may 2016, the dataset was much simpler. 
Since then it looks like it grew almost 6 times in size, most likely upping the complexity.

As a last resort, there is the ["optimization"](https://github.com/findie/wikidata-neo4j-importer/tree/optimization) branch which might be of help. It was never finished as the project was put on hold, but the stage 2 refactor should be stable. It requires APOC to create the relationships dynamically and possibly avoiding some deadlocks.

**Comment by DanielJamieson** [comment]
Thanks @legraphista that makes sense. I'll try the optimisation branch. If that doesn't work, this is a problem I need to solve, so I'll report back with anything I find here.

**Comment by DanielJamieson** [comment]
So on the master branch the error causing the deadlock is this: `Neo.ClientError.Transaction.TransactionHookFailed` (with no message).

On the optimization branch I get this one: `Neo.ClientError.Schema.ConstraintValidationFailed` paired with the message: 
`Node(54951621) already exists with label 'Claim' and property 'id' = \'Q59055$948C6DB7-6F37-41C4-B094-8815EC578583\'`

So it looks like the idempotency isn't working as you'd expect. This could be partly caused by the fact that err.signature in the error catching block actually returns undefined. You can retrieve the code via err.fields[0].code for error catching however. 

I'll see if I can move the process along in the case the claim already exists.

**Comment by DanielJamieson** [comment]
So I've implemented a potential fix. Proper error catching for this 'ConstraintValidationFailed' error such that I can then run `_then` function in that scenario. I'll let this potential fix run through the full dataset and then I'll validate that it has worked properly.

I'm a little concerned that this approach will not insert all claims correctly, as if an error is thrown on item 492 of 1000 items (with bucket size = 1000) and the apoc create relationship statement doesn't try to insert the remaining 508 items, then we will be missing data. If you have knowledge which confirms this, let me know. Otherwise I will validate it myself after the script has finished running, using some data from the last line of each bucket and logging when this error handling is used.

**Comment by DanielJamieson** [comment]
So two things:
a) My process stopped at some point during the script - not clear why but I believe it was killed due to being out of memory. I'll therefore upgrade to 16GB RAM/6 CPUs, which definitely meets the hardware specs in the readme.
b) After having tried to verify whether lines at the end of the bucket were successfully inserted into the database (see previous comment), it looks like it has not been. Therefore I've had to dive a little deeper into the transaction command, and I believe I have some modifications which should allow the whole bucket to be executed while guaranteeing idempotency. 

I'll set the script running again once the VPS has finished upgrading, and I'll let you know how it all goes.

**Comment by DanielJamieson** [comment]
So Stage 2 is finished now and works with the changes I made (concurrency is still at 1 so it was quite slow). Through these changes I think I've realised that the original deadlock problem was not a problem with your code, but an issue when combined with my use of [neo4j-uuid plugin](https://github.com/graphaware/neo4j-uuid).

I think the original error was caused by line 101 in [stage-2/index.js](https://github.com/findie/wikidata-neo4j-importer/blob/master/stage-2/index.js) - which looks like `end = claim.node`. This line replaces the properties of `end` with the properties contained in the `claim.node` structure. When using the neo4j-uuid plugin each node has an additional 'uuid' field which cannot be replaced. It's a little hard to understand exactly if this is the line causing issues, my knowledge of debugging neo4j transaction statements is not great yet, and I can only see that the transaction throws the `Neo.ClientError.Transaction.TransactionHookFailed` error. However, in the optimization branch this statement is on [line 81](https://github.com/findie/wikidata-neo4j-importer/blob/optimization/stage-2/index.js), and removing this line stops the `TransactionHookFailed` error, and makes it work. So I believe this line was the source of the original problem on master branch when combined with neo4j-uuid. I updated it to:

`end += claim.node` 

This simply adds properties from claim.node to end, without trying to remove the uuid property. Again, I have not confirmed that this was the problem on the master branch, but that is my current thinking.

I am not sure whether this change from `=` to `+=` is desirable for your software? Would love to hear your thoughts. You can read about the difference between these two operators [here](https://neo4j.com/docs/cypher-manual/3.5/clauses/set/#set-setting-properties-using-map). 

Even if you don't want the above change, there are still some changes which would be good to include in a potential pull request, at least for the optimization branch. These are:
1) The new error handling for the deadlock retrier, which accesses the error message code incorrectly (I think this must have worked fine for an older neo4j version, but neo4j has updated the error message structure since then and so it doesn't work with newer versions)
2) Replacement of "CREATE" with "MERGE" on line 79 of [stage-2/index.js](https://github.com/findie/wikidata-neo4j-importer/blob/optimization/stage-2/index.js) - this ensures idempotency.

Stage-3 doesn't seem to be working or is very silent for me right now. I'll let that run for longer to see if it does complete. However it isn't too important to me, and you never mentioned this stage being stable in optimization branch. 

Let me know if you'd like me to create a pull request for this - and if yes then whether you'd like the `+=` included too.

**Comment by legraphista** [comment]
Hi Daniel,
First of all, thank you for the work and time you put into finding the issue here.
___
When writing the queries, i did not expect nodes to be able to have immutable properties.

>I am not sure whether this change from = to += is desirable for your software? Would love to hear your thoughts. You can read about the difference between these two operators here.

I agree, moving from `=` to `+=` would be more desirable for such cases, although i'm not familiar with the performance penalties that may occur.
___
>1. The new error handling for the deadlock retrier, which accesses the error message code incorrectly (I think this must have worked fine for an older neo4j version, but neo4j has updated the error message structure since then and so it doesn't work with newer versions)

I fully agree with updating this part of the system, as it was a hack to begin with. I don't remember of any better way (when the code was implemented) to check for deadlocks unless you looked at the code of the result.
___
>2. Replacement of "CREATE" with "MERGE" on line 79 of stage-2/index.js - this ensures idempotency.

Yes, you are right here, my only concern is performance again.
From some testing on the earlier 3.x versions of Neo4j, I noticed that there were cases where if you started 2 `MERGE` queries simultaneously based on the same criteria and if the node didn't exist already, it would create 2 identical copies with the same criteria (unless you had a constraint). This happened because both checks for the node existence would complete before any creation actually occurred, thus both taking the create action instead of update.
I'm not sure if this behaviour still applicable now.
___
>Stage-3 doesn't seem to be working or is very silent for me right now. I'll let that run for longer to see if it does complete. However it isn't too important to me, and you never mentioned this stage being stable in optimization branch.

Yes, stage 3 is a bit silent, it is there only to link some special properties
https://github.com/findie/wikidata-neo4j-importer/blob/36183f5b627c31a9eed826e888369aec5d23d5a2/stage-3/index.js#L21-L30
https://github.com/findie/wikidata-neo4j-importer/blob/36183f5b627c31a9eed826e888369aec5d23d5a2/stage-3/index.js#L44-L53
___
> Let me know if you'd like me to create a pull request for this - and if yes then whether you'd like the += included too.

We would love a pull request from you! ❤️ 
Great work!

**Comment by DanielJamieson** [comment]
Much appreciated @legraphista :) It was a pleasure to contribute, thanks for supporting me along the way! And thanks again for this handy piece of software, it has saved me a lot of time for my own project, so I'm very grateful for its existance.

I've submitted a pull request for the changes mentioned. While testing these changes for the pull request, I discovered that the original error handling in this optimization branch works as it should - It was actually me that forgot to run `npm install` a second time after checking out the optimization branch, so I was using an older verison of the neo4j driver (1.0.1) which the master branch asks for, instead of `^1.2.0`, which the optimization branch wants.

Because the two errors mentioned above "TransactionHookFailed" and "ConstraintValidationFailed" no longer occur when using the changes in the pull request, I found no reason to explicitly catch them in `helper.js`. Therefore my pull request contains only changes to the generate claims transaction.

Thanks again, it was a joy to contribute :)
